xarray/namedarray/core.py

```diff
@@ -847,5 +856,118 @@ def _to_dense(self) -> Self:
         else:
             raise TypeError("self.data is not a sparse array")
 
+    def transpose(
+        self,
+        *dims: Hashable | ellipsis,
+        missing_dims: ErrorOptionsWithWarn = "raise",
+    ) -> Self:
+        """Return a new object with transposed dimensions.
+
+        Parameters
+        ----------
+        *dims : Hashable, optional
+            By default, reverse the dimensions. Otherwise, reorder the
+            dimensions to this order.
+        missing_dims : {"raise", "warn", "ignore"}, default: "raise"
+            What to do if dimensions that should be selected from are not present in the
+            NamedArray:
+            - "raise": raise an exception
+            - "warn": raise a warning, and ignore the missing dimensions
+            - "ignore": ignore the missing dimensions
+
+        Returns
+        -------
+        transposed : NamedArray
+            The returned object has transposed data and dimensions with the
+            same attributes as the original.
+
+        Notes
+        -----
+        This operation returns a view of this variable's data. It is
+        lazy for dask-backed Variables but not for numpy-backed Variables.
+
+        See Also
+        --------
+        numpy.transpose
+        """
+        from xarray.core.indexing import as_indexable
+
+        if not dims:
+            dims = self.dims[::-1]
+        else:
+            dims = tuple(infix_dims(dims, self.dims, missing_dims))
+
+        if len(dims) < 2 or dims == self.dims:
+            # no need to transpose if only one dimension
+            # or dims are in same order
+            return self.copy(deep=False)
+
+        axes = self.get_axis_num(dims)
+        data = as_indexable(self._data).transpose(axes)  # type: ignore
+        return self._replace(dims=dims, data=data)
+
+    @property
+    def T(self) -> Self:
+        return self.transpose()
+
+    def broadcast_to(self, shape: _ShapeLike) -> duckarray[Any, Any]:
+        from xarray.core import duck_array_ops  # TODO: remove this import
+
+        return duck_array_ops.broadcast_to(self.data, shape)  # type: ignore
```

Comment:
do we want this to return a bare duckarray or a `NamedArray` instance??

Reply:
always NamedArray! which means we'll need a dictionary as input `{dim_name: size}`

Reply:
`broadcast_to` is an array_api function if you want to start from that direction.

https://data-apis.org/array-api/latest/API_specification/generated/array_api.broadcast_to.html

Reply:
Done!